File: klu erp try/gateway-python/app/services/satire/satire_engine.py
"""
Satirical Transformation Engine
Orchestrates LLM / Heuristic rewrites of real-world search items and news
into GTA 6 Vice City / Leonida Weazel News satirical feeds.
"""

import json
import random
import httpx
from typing import Dict, List, Any, Optional

from app.config import settings
from app.utils.logger import logger
from app.utils.sanitizer import sanitize_text, sanitize_float
from app.services.satire.lore_dictionary import translate_lore
from app.services.satire.prompts import SATIRE_SYSTEM_PROMPT, build_satire_user_prompt

POSITIVE_SENTIMENT_KEYWORDS = [
    "profit", "surge", "record", "growth", "breakthrough", "rally", "gain",
    "launch", "billion", "merger", "boom", "success", "expand", "soar", "dividend", "jump"
]

NEGATIVE_SENTIMENT_KEYWORDS = [
    "crash", "plunge", "fraud", "lawsuit", "arrest", "scandal", "hack",
    "leak", "investigation", "recall", "layoff", "bankrupt", "crisis", "drop", "ban",
    "disaster", "catastrophic"
]

SATIRICAL_PREFIXES = [
    "Weazel News Exclusive: ",
    "Leonida Man Update: ",
    "BAWSAQ Market Alert: ",
    "Vice City Police Blotter: ",
    "Bleeter Viral Sensation: ",
    "FIB Leaked Memo: ",
    "Consumer Warning: ",
]

SATIRICAL_AUTHORS = [
    "Weazel News Wire",
    "Vice City Inquirer",
    "Bleeter Trending Desk",
    "BAWSAQ Market Desk",
    "VCPD Public Affairs Office",
    "Leonida Consumer Protection Coalition",
]


def calculate_market_impact(text: str) -> float:
    """
    Compute BAWSAQ market impact score (-1.0 to 1.0) based on keyword sentiment
    """
    if not text or not isinstance(text, str):
        return 0.0

    text_lower = text.lower()
    score = 0.0
    for kw in POSITIVE_SENTIMENT_KEYWORDS:
        if kw in text_lower:
            score += 0.25
    for kw in NEGATIVE_SENTIMENT_KEYWORDS:
        if kw in text_lower:
            score -= 0.35

    salt = (len(text) % 10) * 0.02 - 0.09
    total = score + salt if score != 0.0 else salt

    return float(sanitize_float(round(total, 2), default=0.0, min_val=-1.0, max_val=1.0))


class SatireEngine:
    def __init__(self):
        self.provider = settings.satire_provider.lower()

    async def satirize_batch(
        self, items: List[Dict[str, Any]], query: str = "", category: str = "web"
    ) -> List[Dict[str, Any]]:
        """
        Process a list of raw search results into satirical Vice City items
        """
        if not items:
            return []

        results = []
        for item in items:
            satirical = await self.transform_content(item, query, category)
            results.append(satirical)

        return results

    async def transform_content(
        self, item: Dict[str, Any], query: str = "", category: str = "web"
    ) -> Dict[str, Any]:
        """
        Transform a single raw item into GTA satire using configured provider or heuristic fallback
        """
        raw_title = item.get("title") or item.get("headline") or "Breaking News"
        raw_snippet = item.get("snippet") or item.get("body") or "No further details available."
        source_url = item.get("source_url") or item.get("url") or ""
        thumbnail_url = item.get("thumbnail_url") or ""

        # 1. Try LLM Provider if configured and key is present
        if self.provider == "openai" and settings.openai_api_key:
            try:
                transformed = await self._transform_openai(raw_title, raw_snippet, query, category)
                if transformed:
                    return self._format_result(transformed, source_url, thumbnail_url)
            except Exception as e:
                logger.warn(f"OpenAI satire generation failed: {e}. Using heuristic fallback.")

        elif self.provider == "ollama":
            try:
                transformed = await self._transform_ollama(raw_title, raw_snippet, query, category)
                if transformed:
                    return self._format_result(transformed, source_url, thumbnail_url)
            except Exception as e:
                logger.warn(f"Ollama satire generation failed: {e}. Using heuristic fallback.")

        # 2. Ultra-fast, deterministic rule-based Heuristic Transformer (Guaranteed 0ms latency)
        heuristic = self._transform_heuristic(raw_title, raw_snippet, query, category)
        return self._format_result(heuristic, source_url, thumbnail_url)

    def _transform_heuristic(
        self, raw_title: str, raw_snippet: str, query: str = "", category: str = "web"
    ) -> Dict[str, Any]:
        """
        Rule-based heuristic transformer translating real entities and generating satirical commentary
        """
        translated_title = translate_lore(raw_title)
        translated_snippet = translate_lore(raw_snippet)

        # Calculate sentiment and market impact score
        combined_text = f"{raw_title} {raw_snippet}".lower()
        impact = calculate_market_impact(combined_text)

        # Determine satirical author
        author = random.choice(SATIRICAL_AUTHORS)
        if "stock" in category or "market" in combined_text or "finance" in combined_text:
            author = "BAWSAQ Market Desk"
        elif "crime" in category or "police" in combined_text or "arrest" in combined_text:
            author = "VCPD Blotter Desk"
        elif "tech" in category or "phone" in combined_text or "crypto" in combined_text:
            author = "Eyefind Tech Review"

        # Apply satirical headline embellishment if not already present
        headline = translated_title
        if not any(headline.startswith(p) for p in SATIRICAL_PREFIXES) and len(headline) < 90:
            prefix = random.choice(SATIRICAL_PREFIXES)
            headline = f"{prefix}{headline}"

        # Construct GTA-style satirical body
        body = translated_snippet
        if len(body) < 150:
            body = f"{body} Local officials in Vice City recommend citizens invest heavily in ammunition and titanium deadbolts."

        return {
            "title": sanitize_text(headline, settings.max_title_length),
            "snippet": sanitize_text(body, settings.max_snippet_length),
            "satirical_author": author,
            "market_impact": impact,
        }

    def _transform_with_heuristics(
        self, item_or_title: Any, raw_snippet: str = "", query: str = "", category: str = "web"
    ) -> Dict[str, Any]:
        """
        Alias supporting dict or positional parameters for heuristic transformation
        """
        if isinstance(item_or_title, dict):
            raw_title = item_or_title.get("title") or item_or_title.get("headline") or "Breaking News"
            raw_snippet = item_or_title.get("snippet") or item_or_title.get("body") or "No further details available."
            category = item_or_title.get("category", category)
            query = item_or_title.get("query", query)
            return self._transform_heuristic(raw_title, raw_snippet, query, category)
        return self._transform_heuristic(item_or_title, raw_snippet, query, category)

    def _calculate_market_impact(self, text: str) -> float:
        """
        Compute BAWSAQ market impact score (-1.0 to 1.0) based on keyword sentiment
        """
        return calculate_market_impact(text)

    def _format_result(
        self, data: Dict[str, Any], source_url: str, thumbnail_url: str
    ) -> Dict[str, Any]:
        return {
            "title": sanitize_text(data.get("title", ""), settings.max_title_length),
            "snippet": sanitize_text(data.get("snippet", ""), settings.max_snippet_length),
            "source_url": source_url,
            "thumbnail_url": thumbnail_url,
            "satirical_author": sanitize_text(data.get("satirical_author", "Weazel News"), 64),
            "market_impact": sanitize_float(data.get("market_impact", 0.0)),
        }

    async def _transform_openai(
        self, title: str, snippet: str, query: str, category: str
    ) -> Optional[Dict[str, Any]]:
        url = "https://api.openai.com/v1/chat/completions"
        headers = {
            "Authorization": f"Bearer {settings.openai_api_key}",
            "Content-Type": "application/json",
        }
        payload = {
            "model": settings.satire_model,
            "messages": [
                {"role": "system", "content": SATIRE_SYSTEM_PROMPT},
                {"role": "user", "content": build_satire_user_prompt(title, snippet, query, category)},
            ],
            "response_format": {"type": "json_object"},
            "temperature": 0.8,
            "max_tokens": 256,
        }
        async with httpx.AsyncClient(timeout=4.0) as client:
            resp = await client.post(url, headers=headers, json=payload)
            if resp.status_code == 200:
                data = resp.json()
                content = data["choices"][0]["message"]["content"]
                return json.loads(content)
        return None

    async def _transform_ollama(
        self, title: str, snippet: str, query: str, category: str
    ) -> Optional[Dict[str, Any]]:
        url = f"{settings.ollama_host}/api/generate"
        prompt = f"{SATIRE_SYSTEM_PROMPT}\n\n{build_satire_user_prompt(title, snippet, query, category)}"
        payload = {
            "model": settings.ollama_model,
            "prompt": prompt,
            "format": "json",
            "stream": False,
        }
        async with httpx.AsyncClient(timeout=4.0) as client:
            resp = await client.post(url, json=payload)
            if resp.status_code == 200:
                data = resp.json()
                return json.loads(data.get("response", "{}"))
        return None


satire_engine = SatireEngine()

File: klu erp try/gateway-node/src/services/satire/satire.engine.js
/**
 * Satirical LLM Ingestion Layer
 * Transforms real-world search results and news into biting Vice City / Leonida satire.
 * Supports OpenAI, Anthropic, Ollama, with high-velocity heuristic rule-based engine fallback.
 */

const config = require('../../config');
const logger = require('../../utils/logger');
const { LORE_MAPPINGS, SATIRICAL_AUTHORS, SATIRICAL_TEMPLATES } = require('./lore.dictionary');
const { SYSTEM_PROMPT, buildSearchTransformationPrompt } = require('./prompt.templates');
const { sanitizeText } = require('../../utils/sanitizer');

class SatireEngine {
  constructor() {
    this.provider = config.satire.provider;
  }

  /**
   * Main entry point to satirize a single search item
   * @param {Object} item - { title, snippet, source_url, thumbnail_url }
   * @param {string} query - user query
   * @param {string} category - search category
   * @returns {Promise<Object>} Satirized item
   */
  async satirizeSearchItem(item, query = '', category = 'web') {
    // If provider is set to an LLM and API key exists, try LLM first
    if (this.provider === 'openai' && config.satire.openaiApiKey) {
      try {
        const result = await this._transformWithOpenAI(item.title, item.snippet, query, category);
        if (result) return this._mergeResult(item, result);
      } catch (err) {
        logger.warn(`OpenAI satire generation failed, falling back to heuristic: ${err.message}`);
      }
    } else if (this.provider === 'anthropic' && config.satire.anthropicApiKey) {
      try {
        const result = await this._transformWithAnthropic(item.title, item.snippet, query, category);
        if (result) return this._mergeResult(item, result);
      } catch (err) {
        logger.warn(`Anthropic satire generation failed, falling back to heuristic: ${err.message}`);
      }
    } else if (this.provider === 'ollama') {
      try {
        const result = await this._transformWithOllama(item.title, item.snippet, query, category);
        if (result) return this._mergeResult(item, result);
      } catch (err) {
        logger.warn(`Ollama satire generation failed, falling back to heuristic: ${err.message}`);
      }
    }

    // Default High-Velocity Heuristic Satire Engine
    return this._transformHeuristic(item, query, category);
  }

  /**
   * Batch satirize multiple items concurrently
   */
  async satirizeBatch(items, query = '', category = 'web') {
    return Promise.all(items.map((item) => this.satirizeSearchItem(item, query, category)));
  }

  /**
   * Heuristic Rule-Based Satire Engine (Zero latency, 100% offline reliable)
   */
  _transformHeuristic(item, query = '', category = 'web') {
    let title = item.title || '';
    let snippet = item.snippet || '';

    // 1. Apply Lore Replacements across all categories
    for (const group of Object.values(LORE_MAPPINGS)) {
      for (const mapping of group) {
        title = title.replace(mapping.pattern, mapping.replacement);
        snippet = snippet.replace(mapping.pattern, mapping.replacement);
      }
    }

    // 2. Inject Vice City Satire Framing
    const hash = this._simpleHash(title + query);
    const punchIndex = hash % SATIRICAL_TEMPLATES.satiricalPunches.length;
    const suffixIndex = (hash >> 2) % SATIRICAL_TEMPLATES.newsSuffixes.length;
    const authorIndex = (hash >> 4) % SATIRICAL_AUTHORS.length;

    // Satirize title with punch if not already heavily modified
    if (!title.toLowerCase().includes('vice city') && !title.toLowerCase().includes('weazel')) {
      const punchPrefixes = [
        'WEAZEL EXCLUSIVE: ',
        'BREAKING FROM VICE CITY: ',
        'BAWSAQ ALERT: ',
        'LEONIDA MAN REPORT: ',
        'LIFEINVADER LEAK: ',
      ];
      const prefix = punchPrefixes[hash % punchPrefixes.length];
      title = `${prefix}${title}`;
    }

    // Append satirical ending to snippet
    snippet = `${snippet}${SATIRICAL_TEMPLATES.newsSuffixes[suffixIndex]}`;

    // 3. Compute Market Impact (-1.0 to 1.0)
    const marketImpact = this._calculateMarketImpact(title + ' ' + snippet, hash);

    const author = SATIRICAL_AUTHORS[authorIndex];

    return {
      title: sanitizeText(title, config.safety.maxTitleLength),
      snippet: sanitizeText(snippet, config.safety.maxSnippetLength),
      source_url: item.source_url || '',
      thumbnail_url: item.thumbnail_url || '',
      satirical_author: author,
      market_impact: marketImpact,
    };
  }

  /**
   * Compute in-game market sentiment score based on key financial & comedic triggers
   */
  _calculateMarketImpact(text, seed) {
    const lower = text.toLowerCase();
    let score = 0.0;

    // Positive market triggers
    if (lower.includes('profit') || lower.includes('surge') || lower.includes('record') || lower.includes('monopoly') || lower.includes('deregulation') || lower.includes('tax evasion')) {
      score += 0.45;
    }
    if (lower.includes('subscribers') || lower.includes('unveils') || lower.includes('billion') || lower.includes('expansion')) {
      score += 0.3;
    }

    // Negative market triggers
    if (lower.includes('crash') || lower.includes('fraud') || lower.includes('investigation') || lower.includes('fib') || lower.includes('raid') || lower.includes('lawsuit')) {
      score -= 0.55;
    }
    if (lower.includes('explosion') || lower.includes('boycott') || lower.includes('banned') || lower.includes('scandal')) {
      score -= 0.35;
    }

    // Add deterministic micro-variance
    const variance = ((seed % 100) - 50) / 250.0;
    score = Math.max(-1.0, Math.min(1.0, score + variance));

    return parseFloat(score.toFixed(2));
  }

  /**
   * Simple string hash for deterministic randomized selection
   */
  _simpleHash(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      hash = (hash << 5) - hash + str.charCodeAt(i);
      hash |= 0;
    }
    return Math.abs(hash);
  }

  /**
   * OpenAI API Integration
   */
  async _transformWithOpenAI(title, snippet, query, category) {
    const prompt = buildSearchTransformationPrompt(title, snippet, query, category);
    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${config.satire.openaiApiKey}`,
      },
      body: JSON.stringify({
        model: config.satire.openaiModel,
        messages: [
          { role: 'system', content: SYSTEM_PROMPT },
          { role: 'user', content: prompt },
        ],
        temperature: config.satire.temperature,
        response_format: { type: 'json_object' },
      }),
    });

    if (!response.ok) {
      throw new Error(`OpenAI HTTP ${response.status}: ${await response.text()}`);
    }

    const data = await response.json();
    const content = data.choices?.[0]?.message?.content;
    return JSON.parse(content);
  }

  /**
   * Anthropic API Integration
   */
  async _transformWithAnthropic(title, snippet, query, category) {
    const prompt = buildSearchTransformationPrompt(title, snippet, query, category);
    const response = await fetch('https://api.anthropic.com/v1/messages', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': config.satire.anthropicApiKey,
        'anthropic-version': '2023-06-01',
      },
      body: JSON.stringify({
        model: config.satire.anthropicModel,
        system: SYSTEM_PROMPT,
        messages: [{ role: 'user', content: prompt }],
        max_tokens: 500,
        temperature: config.satire.temperature,
      }),
    });

    if (!response.ok) {
      throw new Error(`Anthropic HTTP ${response.status}: ${await response.text()}`);
    }

    const data = await response.json();
    const text = data.content?.[0]?.text;
    const jsonMatch = text.match(/\{[\s\S]*\}/);
    if (jsonMatch) {
      return JSON.parse(jsonMatch[0]);
    }
    throw new Error('Failed to parse JSON from Anthropic response');
  }

  /**
   * Ollama Local LLM Integration
   */
  async _transformWithOllama(title, snippet, query, category) {
    const prompt = buildSearchTransformationPrompt(title, snippet, query, category);
    const response = await fetch(`${config.satire.ollamaBaseUrl}/api/generate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model: config.satire.ollamaModel,
        system: SYSTEM_PROMPT,
        prompt: prompt,
        format: 'json',
        stream: false,
      }),
    });

    if (!response.ok) {
      throw new Error(`Ollama HTTP ${response.status}`);
    }

    const data = await response.json();
    return JSON.parse(data.response);
  }

  _mergeResult(original, transformed) {
    return {
      title: sanitizeText(transformed.title || original.title, config.safety.maxTitleLength),
      snippet: sanitizeText(transformed.snippet || original.snippet, config.safety.maxSnippetLength),
      source_url: original.source_url || '',
      thumbnail_url: original.thumbnail_url || '',
      satirical_author: sanitizeText(transformed.satirical_author || 'Weazel News Correspondent', 64),
      market_impact: typeof transformed.market_impact === 'number' ? Math.max(-1.0, Math.min(1.0, transformed.market_impact)) : 0.0,
    };
  }
}

const satireEngine = new SatireEngine();

function calculateMarketImpact(text, seed = 42) {
  return satireEngine._calculateMarketImpact(text, seed);
}

function transformWithHeuristics(item, query = '', category = 'web') {
  return satireEngine._transformHeuristic(item, query, category);
}

async function transformContent(item, query = '', category = 'web') {
  return satireEngine.satirizeSearchItem(item, query, category);
}

satireEngine.calculateMarketImpact = calculateMarketImpact;
satireEngine.transformWithHeuristics = transformWithHeuristics;
satireEngine.transformContent = transformContent;

module.exports = satireEngine;
module.exports.calculateMarketImpact = calculateMarketImpact;
module.exports.transformWithHeuristics = transformWithHeuristics;
module.exports.transformContent = transformContent;
